src/scheduler/job_info.c

```diff
@@ -2964,25 +2968,75 @@ find_and_preempt_jobs(status *policy, int pbs_sd, resource_resv *hjob, server_in
 	 * or the maximum number of tries has been exhausted
 	 */
 	while (!done &&
-		((jobs = find_jobs_to_preempt(policy, hjob, sinfo, fail_list)) != NULL) &&
+		((jobs = find_jobs_to_preempt(policy, hjob, sinfo, fail_list, &count)) != NULL) &&
 		num_tries < MAX_PREEMPT_RETRIES) {
 		done = 1;
-		for (i = 0 ; jobs[i] != 0; i++) {
+
+		if ((preempt_jobs_list = calloc(count+1, sizeof(char *))) == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
+			free(preempted_list);
+			free(fail_list);
+			return -1;
+		}
+
+		for (i = 0; i < count; i++) {
 			job = find_resource_resv_by_indrank(sinfo->running_jobs, jobs[i], -1);
 			if (job != NULL) {
-				ret = preempt_job(policy, pbs_sd, job, sinfo);
-
-				if (ret)
-					/* copy this job into the preempted array list */
-					preempted_list[preempted_count++] = jobs[i];
-				else  {
-					done = 0; /* preemption failed for some job, need to loop */
-					/* add to the fail list */
-					fail_list[fail_count++] = jobs[i];
+				if ((preempt_jobs_list[i] = calloc(PBS_MAXSVRJOBID+1, 1)) == NULL) {
+					log_err(errno, __func__, MEM_ERR_MSG);
+					for (j = 0; j < i; j++)
+						free(preempt_jobs_list[j]);
+					free(preempt_jobs_list);
+					free(preempted_list);
+					free(fail_list);
+					return -1;
+				}
+				snprintf(preempt_jobs_list[i], PBS_MAXSVRJOBID, "%s", job->name);
+			}
+		}
+
+		preempt_jobs_reply = pbs_preempt_jobs(pbs_sd, preempt_jobs_list);
+
+		for (i = 0; i < count; i++) {
+			job = find_resource_resv(sinfo->running_jobs, preempt_jobs_reply[i].job_id);
+			if (preempt_jobs_reply[i].order[0] == '0')
+				fail_list[fail_count++] = job->rank;
+			else {
+				preempted_list[preempted_count++] = job->rank;
+				if (preempt_jobs_reply[i].order[0] == 'S') {
+					/* Set resources_released and execselect on the job */
+					create_res_released(policy, job);
+
+					update_universe_on_end(policy, job, "S", NO_FLAGS);
+					job->job->is_susp_sched = 1;
+					schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_INFO,
+						job->name, "Job preempted by suspension");
+				} else if (preempt_jobs_reply[i].order[0] == 'C') {
+					job->job->is_checkpointed = 1;
+					update_universe_on_end(policy, job, "Q", NO_FLAGS);
+					schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_INFO,
+						job->name, "Job preempted by checkpointing");
+					//if (pbs_sd != SIMULATE_SD)
+						//pbs_statfree(status);
+
+					pbs_rlsjob(pbs_sd, job->name, "s", NULL);
+				} else {
+					update_universe_on_end(policy, job, "Q", NO_FLAGS);
+					schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_INFO,
+						job->name, "Job preempted by requeuing");
 				}
+				update_accruetype(pbs_sd, sinfo, ACCRUE_MAKE_ELIGIBLE, SUCCESS, job);
+				mark_job_preempted(pbs_sd, job, sinfo->server_time);
```

Comment:
is_preempted is being used in the scheduler, so keeping it here for now.

Reply:
The scheduler queries the preempted attribute every cycle.  I don't see any purpose to having the scheduler set it.  The server is the entity doing the preemption, so why isn't setting the attribute.  It saves the back and forth from scheduler to the server.  You'll still need to flip the scheduler's is_preempted bit on here.

Reply:
As I see scheduler uses this attribute and server just keeps it to send it to the scheduler when the scheduler queries the jobs.
If I am not missing anything, the current sequence of events is -
1) query the jobs - if this attribute is set, we set time_preempted and is_preempted for the job.
2) send preempt jobs request to the server
3) update the time_preempted and is_preempted for the job(s) that was just preempted, send an update of this attribute to the server.
4) update_preemption_on_run will use the value we just updated (cmp_job_preemption_time_asc).

If we have the server update this attribute - the sequence will be -
1) query the jobs - if this attribute is set, we set time_preempted and is_preempted for the job.
2) send preempt jobs request to the server
3) have the server update this attribute.
4) update_preemption_on_run will **not be able to** use the value we just updated.

Reply:
I'm not sure I understand.  I'm just talking about who sets the attribute on the server.  Right now the scheduler will send a pbs_alterjob() request to the server to set it.  I'm just saying to have the server set it directly.  The scheduler already updates its cache when preempting a job.  It doesn't query the value again.  It can continue updating its cache as it does now.  It just doesn't alter the attribute.